Basic aspect ratio control

DuCorey opened the pull request:
## Summary
This PR adds a simple aspect ratio control tool.

- [x] issues: fixes #474, #517
- [ ] tests 
- [ ] release document entry (if new feature or API change)

## Latest commit
This PR adds a two properties to the plot object.

`match_aspect` which indicates whether Bokeh should attempt the match the (width/height) of *data space* to the (width/height) in pixels of *screen space*.

`aspect_scale` which allows for further control of aspect control behavior by multiplying on top of the ratio calculated at `match_aspect`. This property allows the user to distort the axis at will. For example, with `aspect_scale= 2` a square in *data* units will be drawn as a rectangle with twice the *pixel* units in width than in height.

#### `match_aspect=True`
This fixes the bug shown in #6603.
![bokeh_plot 32](https://user-images.githubusercontent.com/27960786/29531839-645b31ae-8678-11e7-960c-ba3c3b612697.png)

#### `match_aspect=True, aspect_scale=2`
![bokeh_plot 47](https://user-images.githubusercontent.com/27960786/29589097-f268486c-8761-11e7-859c-b125d7812315.png)


## Explanation

Hello Bokeh team,

From my understanding the current behavior when plotting the data ranges is to use the bounds of the glyphs. The method proposed in this PR for aspect control is to instead set the data ranges values with new ones calculated based on the glyph bounds and the inherent aspect ratio of the plot. It also allows an arbitrary user defined aspect ratio. 

### A few examples:
#### Current behavior
![bokeh_plot 42](https://user-images.githubusercontent.com/27960786/29581170-82c13480-8746-11e7-8f26-b7f77ae71ecb.png)

#### New behavior with aspect ratio of 1:
![bokeh_plot 40](https://user-images.githubusercontent.com/27960786/29581229-ab0cbedc-8746-11e7-9311-47ced5276bd1.png)


#### New behavior with aspect ratio of 16/9:
![bokeh_plot 41](https://user-images.githubusercontent.com/27960786/29581189-92f8d466-8746-11e7-80f6-177e1e10aafa.png)


### What it doesn't fix
This isn't a perfect fix, since it simply uses the built in ranges for trying to maintain the aspect ratio. As described in #517, there seems to be a deeper problem with pixel per distance being different on each axis. I'm currently studying the information in those threads to see if my current solution can be improved.

### Todo
- [ ] tests 
- [ ] release document entry (if new feature or API change)

I'm assuming a new feature like this requires tests and unfortunately I haven't used mocha before, but I will do my best to get the tests done.

### Final remarks
I hope this PR is appreciated. It's always of pleasure contributing to a project like Bokeh. Feel free to dismiss this PR if you are working on something more comprehensive that also fixes issues like #6603 and #517. I've also left some more comments on specific parts about the code. I'm welcome to all suggestions or improvements. I'll be available to work some more on the PR after the weekend.

To the whole team, enjoy your weekend and thank you for such a great project.
Corey

DuCorey commented on bokehjs/src/coffee/models/plots/plot_canvas.coffee:
```diff
@@ -210,6 +210,32 @@ export class PlotCanvasView extends DOMView
     has_bounds = false
 
+    if @model.plot.aspect_ratio != false
+      aspect = if @model.plot.aspect_ratio is true then 1 else @model.plot.aspect_ratio
```
Should there be more specific checks for the float number?

DuCorey commented on bokehjs/src/coffee/models/plots/plot_canvas.coffee:
```diff
@@ -210,6 +210,32 @@ export class PlotCanvasView extends DOMView
+      aspect = if @model.plot.aspect_ratio is true then 1 else @model.plot.aspect_ratio
+      r = aspect*(@model.plot.width/@model.plot.height)
+
+      for k, v of bounds
```
I'm still a beginner in Coffescript, but is comprehension the correct way of looping through the object? Would closing the scope be preferred? i.e. 
```
for k, v of bounds 
  do(k, v) -> 
```

DuCorey commented on bokehjs/src/coffee/models/plots/plot_canvas.coffee:
```diff
@@ -210,6 +210,32 @@ export class PlotCanvasView extends DOMView
+        bounds[k].maxX = xcenter+0.5*width
+        bounds[k].minX = xcenter-0.5*width
+        bounds[k].maxY = ycenter+0.5*height
+        bounds[k].minY = ycenter-0.5*height
```
Would it be better to create a new dictionary with new values instead of overwriting bounds?

bryevdv commented:
Whoa, unexpected and awesome

> I hope this PR is appreciated.

An enthusiastic YES. If I may wax slightly, this is the power of OSS. One reason that this issue has lingered for so long is that I had framed the problem in entirely different terms, namely: given fixed/determined data bounds, how can the layout (i.e. the actual pixel dimensions of the central plot area) be adjusted dynamically to maintain some constraints. I imagined lots of bad cases and it made getting started very difficult. But you've come at it from the opposite direction, which never occurred to me at all (taking the pixel bounds as determined and adjusting the data ranges appropriately). But I think this approach here (and limiting things to when there are two data ranges) is actually *much* simpler as well as more promising!

We are in a bit of a crush right before the next release freeze, so my inclination is to let this simmer for the next week or two. That will allow for time to add tests, tweak docs, etc without being rushed. I also think the pixel-distance issue might be fixable with a small delta off this code, so we could potentially discuss and iterate on that as well. Then the goal would be to merge and get it into a dev build soon after `0.12.7` comes out. Does that sound like a reasonable plan of attack?

Again, thank you for the extremely pleasant surprise.

bryevdv commented:
Basically for the pixel distance problem, off hand I might suggest the following: add another property option `iso` (or maybe `auto`) which simply looks at the pixel aspect ratio of the "central" plot frame, and then automatically uses that as the numerical aspect ratio when updating data ranges. 

The reason that per-pixel distances may not match in different directions is that the aspect ration of the pixel dimensions is different from the aspect ratio of the "data" dimensions. But te former is easily obtainable, and you just opened up an easy path to setting the latter.

bryevdv commented on bokehjs/src/coffee/models/plots/plot_canvas.coffee:
```diff
@@ -210,6 +210,32 @@ export class PlotCanvasView extends DOMView
 
+    if @model.plot.aspect_ratio != false
+      aspect = if @model.plot.aspect_ratio is true then 1 else @model.plot.aspect_ratio
+      r = aspect*(@model.plot.width/@model.plot.height)
```
Pretty sure this should be using the plot frame's "inner width" and "inner height", those are the pixel dimensions of the area between the axes, i.e.  not counting any space taken up by title, axes, tick labels, padding, etc.

bryevdv commented:
Another reason to take time is we will need to figure out how things like interactions with tools should behave. If an aspect is set to something specific, and a user tries to zoom just one axis... what happens? Maybe that's fine, and aspect is only applied when autoranging, so if a user explicily changes the bounds (e.g. by zooming) then it's fine to ignore the aspect constraint until a reset. Mostly we just need to make sure we understand the behaviours, and document it well.

bryevdv commented on bokeh/models/plots.py:
```diff
@@ -692,3 +692,10 @@ def webgl(self, webgl):
+    aspect_ratio = Either(Bool, Float, default=False, help="""
+    Specify the aspect ratio behavior of the plot. Default is ``False`` which
+    means that the glyph will fill the whole plot area. ``True`` indicates that
+    the glyphs should maintain the an aspect ratio of 1. A ``Float``value can
```
lack of space after last backtick is what is causing the one test failure (sphinx barfs)

cwhy commented:
I have not tried it yet but why in the "New behavior with aspect ratio of 1" example plot have a aspect ratio that is not exactly 1:1?

bryevdv commented:
> I have not tried it yet but why in the "New behavior with aspect ratio of 1" example plot have a aspect ratio that is not exactly 1:1?

As I commented above, it's using the overall canvas pixel dimensions, instead of the pixel dimensions of just the axes that border the plot area, and I believe that is an issue.

DuCorey commented:
Good news, following @bryevdv suggestions of using the `inner_width` and `inner_height` works perfectly now for setting the aspect ratio. Measuring the pixels on both axis, I'm finding they match up perfectly. I've updated the examples in the PR. 

Second, getting access to the values of `inner_width` and `inner_height` was a little more tricky. From my understanding of the code, the `inner_width` and `inner_height` are set from the `frame._width` and `frame._height` properties which are in turn calculated by `update_constraints()` in the `render()` function of the model. Unfortunately `update_dataranges()` is called in the init before `frame._width` and `frame._height` have been calculated. To go around this, I've added `update_dataranges()` again in the `render()` function. This now properly calculates the range values with aspect ratio. However, from running the code on my side seems to be called multiple times and portions of the `update_dataranges()` are being run unnecessarily. I'm unsure how best to proceed to avoid unnecessary calls to `update_dataranges()`. Let me know what you think.

bryevdv commented:
>  However, from running the code on my side seems to be called multiple times and portions of the update_dataranges() are being run unnecessarily. 

It's possible, given that the data bounds depend on the pixel dims of the axes which in turn depend on the extents of the tick labels which depends on the data bounds that iterating/calling twice may be unavoidable

bryevdv commented:
```
from bokeh.io import output_file, show
from bokeh.plotting import figure

p = figure(aspect_ratio=1, plot_height=400)

p.circle([2], [2], radius=0.4)

show(p)
```
<img width="517" alt="screen shot 2017-08-21 at 13 13 59" src="https://user-images.githubusercontent.com/1078448/29532417-a5b5a784-8672-11e7-9351-b245713c811e.png">


Hooray! 

@DuCorey What do you think about changing `Bool` to `Auto`. Then the default is `None` (`null` on the JS side`) which means don't do anything to set aspect (corresponds to `False` in the current code) and `"auto"` means match the plot area's aspect (corresponds to `True`)?

Also I think I'd have a slight preference for `aspect` instead of `aspect_ratio`

bryevdv commented:
Also this behaves like I expected when zooming just one axis or the other:

<img width="650" alt="screen shot 2017-08-21 at 13 18 42" src="https://user-images.githubusercontent.com/1078448/29532551-4c84aaf6-8673-11e7-8d10-9127239a4961.png">
<img width="649" alt="screen shot 2017-08-21 at 13 18 54" src="https://user-images.githubusercontent.com/1078448/29532550-4c7d1be2-8673-11e7-84b0-d486536e1f1c.png">

This seems fine, we just need to make sure to document that aspect matching only happens on initial auto-ranging and resets. Operations like single axis zoom explicitly will *not* preserve aspect.

bryevdv commented:
lastly! it would be ideal to factor out the code that computes new bounds based on pixel aspect into its own little function and add some tests for it. I can help with that if necessary.

bryevdv commented:
Also regarding re-calling `update_dataranges` that could be skipped if `aspect` is `None` (false)

bryevdv commented:
@DuCorey do you mind if I push changes to your branch? I would like to try and get this into 0.12.7 after all

bryevdv commented:
Alternatively if you have time to make these two changes today:

* rename `aspect_ratio` to `aspect` 

* use `Auto` instead of `Bool`

Then this PR could be merged  and I could add tests and docs in a separate PR. 

I understand if you don't have time on short notice, which is why I am also happy to push changes to your branch directly.

DuCorey commented:
@bryevdv Changes are done. However, even with the added check we are still calling update_dataranges() twice. Going through the code execution, the behavior is consistent because when we call `update_dataranges() `and update the x and y ranges, this calls a request_paint() with `final = false` which then recalculates both the constraints and the updates the ranges however since the ranges aren't being changed this time around the `request_paint()` is sent with` final = true` and the code goes elsewhere. I haven't tried changing this logic.

Second, I've inversed the behavior of aspect to now increase in width when we increase the aspect ratio. To me this makes more sense, since looking at definitions of aspect ratio online we see that for example an aspect ratio of 16:9 indicates a ratio of width over height of 1.77. The method now does the same thing. It sets the ranges such as the physical width over height of plot ranges are equal to 1.77. I think most users using this feature will expect such a behavior.

bryevdv commented:
@DuCorey thanks! It's almost certain we are "over-calling" things in a few different places. There is an open issue about `_map_data` being called more than necessary, e.g. It'll be best to try and investigate and improve these things in their own PR. (Performance generally seems adequate despite over-calling)

> Second, I've inversed the behavior of aspect to now increase in width when we increase the aspect ratio

Yes, good catch, aspect is normally defined by width/height

bryevdv commented:
OK tests are green I will plan on kicking the tires just a bit then merging later this afteroon

bryevdv commented:
Well tests *were* green until I accidentally hit the "close" button I've stopped the new test run, no sense in taking up test workers.

bryevdv commented on bokeh/models/plots.py:
```diff
@@ -692,3 +692,13 @@ def webgl(self, webgl):
+
+    Default is ``None`` which indicates no aspect ratio. ``"auto"`` indicates
+    that the plot should maintain an aspect ratio of 1. A ``Float`` value can
+    be given for arbitrary aspect ratio control.
```
We can edit this in a follow-on PR before the release, but this is a little off. `"auto"` indicates that the aspect ratio of the axes dimensions should be automatically made to match the aspect ratio of the corresponding pixel extents of the axes. The end result is that a 1x1 area in data space is a square in pixels, and conversely that a 1x1 pixel is a square in data units, or (perhaps more importantly) that circles drawn inside squares will always touch all four sides. 

We should also add a note about how this setting only affect when there are two dataranges, and then only affects initial plot draw and resets -- it is possible for the use of tools (single axis zoom or unconstrained bo zoom) to change the aspect ratio

bryevdv commented:
> This shouldn't be merged without image diff or integration tests.

Plan is to add tests and more docs in follow-on PRs during the freeze so that this can go in an RC tomorrow morning. I'm merging when travis goes green.

bryevdv commented:
@canavandl I'm not sure I've ever added a new image diff integration test from scratch. Can we pair on that in the office in the next day or two? Then I will update those docs with more detail in the docs PR. I have asked @DuCorey for an `examples/plotting/file/aspect.py` in a follow on PR that will get picked up by the image diff tests. (I think we need a dedicated integration test too, though)

cc @mattpap

bryevdv commented:
@mattpap Additionally, in any case integration tests will really need to be added by someone on @bokeh/dev in order to make sure they actually function
